### PR TITLE
Add script to build toolchain

### PR DESCRIPTION
**Pull request by johnno1962**
#### What's in this pull request?

A short script to document how to/build toolchain

Tested on OS X/Linux
#### Resolved bug number:

https://bugs.swift.org/browse/SR-174

### CODE REVIEW

**Comment by modocache** [comment]
@johnno1962 I'm curious, as I've tried to do this before: how much disk space does it take to build an OS X toolchain? In my attempts 300GB of free disk space wasn't enough, but maybe my machine's messed up.

**Comment by johnno1962** [comment]
I only have 60GB free. It should take less space than a normal build which has debug symbols:

johnholdsworth$ du -s ../build/buildbot_osx/
31784008	../build/buildbot_osx/
johnholdsworth$ du -s swift-nightly-install/Library/Developer/Toolchains/swift-SNAPSHOT-2016-03-07-a.xctoolchain/
1238592	swift-nightly-install/Library/Developer/Toolchains/swift-SNAPSHOT-2016-03-07-a.xctoolchain/

**Comment by modocache** [comment]
Hmm, must be on my end then--thanks! I'll try again.

File: utils/build-toolchain [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+SWIFT_SOURCE_ROOT="${SRC_DIR}"
+SWIFT_BUILD_ROOT="${SRC_DIR}/build"
+SWIFT_INSTALLABLE_PACKAGE="${SRC_DIR}/${ARCHIVE}"
+SWIFT_INSTALL_DIR="${SRC_DIR}/swift-nightly-install"
+SWIFT_INSTALL_SYMROOT="${SRC_DIR}/swift-nightly-symroot"
+SWIFT_TOOLCHAIN_DIR="/Library/Developer/Toolchains/${TOOLCHAIN_NAME}.xctoolchain"
+SYMBOLS_PACKAGE="${SRC_DIR}/${SYM_ARCHIVE}"
+
+if [ -e /etc/lsb-release ]; then
```
Comment (modocache) [inline utils/build-toolchain]: Is this to determine the host system? If so, `util/build-script-impl` uses `uname -s` for this purpose.

Comment (johnno1962) [inline utils/build-toolchain]: Better:

if [[ "$(uname -s)" == "Darwin" ]] ; then
    SWIFT_PACKAGE=buildbot_osx_package
else
    SWIFT_PACKAGE=buildbot_linux
fi

?

Comment (modocache) [inline utils/build-toolchain]: Personally I think so.

**Comment by johnno1962** [comment]
Thanks @gottesmm

**Comment by johnno1962** [comment]
@swift-ci Please test

**Comment by tkremenek** [comment]
@swift-ci Please test

**Comment by tkremenek** [comment]
@shahmishal Can you take a look at this PR?

File: utils/build-toolchain [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+SWIFT_TOOLCHAIN_DIR="/Library/Developer/Toolchains/${TOOLCHAIN_NAME}.xctoolchain"
+SYMBOLS_PACKAGE="${SRC_DIR}/${SYM_ARCHIVE}"
+
+if [[ "$(uname -s)" == "Darwin" ]] ; then
+    SWIFT_PACKAGE=buildbot_osx_package
+else
+    SWIFT_PACKAGE=buildbot_linux
+fi
+
+./utils/build-script --preset="${SWIFT_PACKAGE}" install_destdir="${SWIFT_INSTALL_DIR}" installable_package="${SWIFT_INSTALLABLE_PACKAGE}" install_toolchain_dir="${SWIFT_TOOLCHAIN_DIR}" install_symroot="${SWIFT_INSTALL_SYMROOT}" symbols_package="${SYMBOLS_PACKAGE}" darwin_toolchain_bundle_identifier="${BUNDLE_IDENTIFIER}" darwin_toolchain_display_name="${DISPLAY_NAME}" darwin_toolchain_xctoolchain_name="${TOOLCHAIN_NAME}" darwin_toolchain_version="${TOOLCHAIN_VERSION}" darwin_toolchain_alias=Local
```
Comment (shahmishal) [inline utils/build-toolchain]: Can you please split up this line so its easy to read? 
```
./utils/build-script --preset="${SWIFT_PACKAGE}" \
    install_destdir="${SWIFT_INSTALL_DIR}" \
    installable_package="${SWIFT_INSTALLABLE_PACKAGE}" \
    install_toolchain_dir="${SWIFT_TOOLCHAIN_DIR}" \
    install_symroot="${SWIFT_INSTALL_SYMROOT}" \
    symbols_package="${SYMBOLS_PACKAGE}" \
    darwin_toolchain_bundle_identifier="${BUNDLE_IDENTIFIER}" \
    darwin_toolchain_display_name="${DISPLAY_NAME}" \
    darwin_toolchain_xctoolchain_name="${TOOLCHAIN_NAME}" \
    darwin_toolchain_version="${TOOLCHAIN_VERSION}" \
    darwin_toolchain_alias=Local
```

File: utils/build-toolchain [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+cd "$(dirname $0)/.."
+SRC_DIR=$PWD
+
+YEAR=$(date +"%Y")
+MONTH=$(date +"%m")
+DAY=$(date +"%d")
+TOOLCHAIN_VERSION="swift-SNAPSHOT-${YEAR}-${MONTH}-${DAY}-a"
+ARCHIVE_DIR="${TOOLCHAIN_VERSION}-${BUILD_NUMBER}"
```
Comment (shahmishal) [inline utils/build-toolchain]: We can delete this `ARCHIVE_DIR`, does not look like this script is using.

**Comment by johnno1962** [comment]
@shahmishal your suggestions submitted, let me know if you want me to submit to master as well

File: utils/build-toolchain [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+cd "$(dirname $0)/.."
+SRC_DIR=$PWD
+
+YEAR=$(date +"%Y")
+MONTH=$(date +"%m")
+DAY=$(date +"%d")
+TOOLCHAIN_VERSION="swift-SNAPSHOT-${YEAR}-${MONTH}-${DAY}-a"
+ARCHIVE="${TOOLCHAIN_VERSION}-osx.tar.gz"
+SYM_ARCHIVE="${TOOLCHAIN_VERSION}-osx-symbols.tar.gz"
+BUNDLE_IDENTIFIER="org.swift.${YEAR}${MONTH}${DAY}"
```
Comment (gribozavr) [inline utils/build-toolchain]: I don't think we should encourage other developers to build toolchain with the "org.swift" identifier.  Only the official CI should be using it.  This part should be a command-line parameter.

Comment (shahmishal) [inline utils/build-toolchain]: +1

**Comment by johnno1962** [comment]
I've added bundle prefix as an argument and changed toolchain install name from SNAPSHOT to LOCAL to avoid overwriting official releases

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@johnno1962 We should commit new changes to master first, and cherry pick them to release branch. Can you please also create PR for `master` branch? Thanks!

**Comment by johnno1962** [comment]
Thanks @shahmishal, I’ve created https://github.com/apple/swift/pull/1608, let me know if there is anything else.

**Comment by johnno1962** [comment]
@shahmishal, I’ve resubmitted a new master version of this PR on a fresh clone so as not to inlcude extra merges: https://github.com/apple/swift/pull/1630. https://github.com/apple/swift/pull/1608 can now be closed.

**Pull request merged**